v2/stream/new/client.go

```diff
@@ -0,0 +1,336 @@
+package new
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+)
+
+// StreamV2Client is a client that connects to an Alpaca Data V2 stream server
+// and handles communication both ways.
+//
+// After constructing, Connect() must be called before any subscription changes
+// are called. Connect keeps the connection alive and reestablishes it until
+// a configured number of retries has not been exceeded.
+//
+// Error() returns a channel that the client sends an error to when it has terminated.
+// A client can not be reused once it has terminated!
+//
+// SubscribeTo... and UbsubscribeFrom... can be used to modify subscriptions and
+// the handler used to process incoming trades/quotes/bars. These block until an
+// irrecoverable error occurs or if they succeed.
+//
+// Note that subscription changes can not be called concurrently.
+type StreamV2Client interface {
+	Connect(ctx context.Context) error
+	Error() <-chan error
+	SubscribeToTrades(handler func(trade Trade), symbols ...string) error
+	UnsubscribeFromTrades(symbols ...string) error
+	SubscribeToQuotes(handler func(quote Quote), symbols ...string) error
+	UnsubscribeFromQuotes(symbols ...string) error
+	SubscribeToBars(handler func(bar Bar), symbols ...string) error
+	UnsubscribeFromBars(symbols ...string) error
+}
+
+type client struct {
+	logger Logger
+
+	feed   string
+	host   string
+	key    string
+	secret string
+
+	reconnectLimit int
+	reconnectDelay time.Duration
+	processorCount int
+	connectOnce    sync.Once
+	connectCalled  bool
+	hasTerminated  bool
+	connErrChan    chan error
+	conn           conn
+	in             chan []byte
+	subChanges     chan []byte
+
+	handlerMutex          sync.RWMutex
+	trades                []string
+	tradeHandler          func(trade Trade)
+	quotes                []string
+	quoteHandler          func(quote Quote)
+	bars                  []string
+	barHandler            func(bar Bar)
+	pendingSubChangeMutex sync.Mutex
+	pendingSubChange      *subChangeRequest
+}
+
+var _ StreamV2Client = (*client)(nil)
+
+func NewClient(feed string, opts ...Option) StreamV2Client {
+	c := client{
+		feed:        feed,
+		connErrChan: make(chan error, 1),
+		subChanges:  make(chan []byte, 1),
+	}
+	o := defaultOptions()
+	o.apply(opts...)
+	c.configure(*o)
+	return &c
+}
+
+var connCreator = func(ctx context.Context, u url.URL) (conn, error) {
+	return newNhooyrWebsocketConn(ctx, u)
+}
+
+func constructURL(host, feed string) (url.URL, error) {
+	scheme := "wss"
+	ub, err := url.Parse(host)
+	if err != nil {
+		return url.URL{}, err
+	}
+	switch ub.Scheme {
+	case "http", "ws":
+		scheme = "ws"
+	}
+
+	return url.URL{Scheme: scheme, Host: ub.Host, Path: "/v2/" + feed}, nil
+}
+
+var ErrConnectCalledMultipleTimes = errors.New("Connect called multiple times")
+
+// Connect establishes a connection for c using its configuration and reestablishes it when errors occur.
```

Comment:
Somehow I would try to highlight that it also keeps the connection open in the background forever.

Reply:
I'll give it a shot.